### PR TITLE
Improve logging tests isolation

### PR DESCRIPTION
**Pull request by taburet**
Provides a new JUnit test rule for use by logging related tests.
The rule (IsolatedLoggingRule) fully isolates the logging state
from other tests.

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/test/IsolatedLoggingRule.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+import java.lang.reflect.Field;
+
+/**
+ * Fully isolates the shared global logging state for a test. Also provides
+ * utilities which are useful while testing logging related things.
+ * <p>
+ * Make sure you are using a serial test runner if you use this rule.
```
Comment (blazember) [inline hazelcast/src/test/java/com/hazelcast/test/IsolatedLoggingRule.java]: I think we can even enforce it. `Description` param of `apply` seems to know about the test and the test class being executed. 

Adding something like the following to the beginning of the `apply` method below may do the trick.
```
        // making sure that there are no concurrent test executions so we don't impact their results unintentionally
        RunWith runWith = description.getTestClass().getAnnotation(RunWith.class);
        if (!HazelcastSerialClassRunner.class.equals(runWith.value())) {
            throw new IllegalStateException("Tests using this rule must use serial runner");
        }
```

Comment (taburet) [reply hazelcast/src/test/java/com/hazelcast/test/IsolatedLoggingRule.java]: Good idea, will definitely incorporate this into the fix.

**Comment by Donnerbart** [comment]
run-lab-run

**Comment by taburet** [comment]
run-lab-run

File: hazelcast/src/test/java/com/hazelcast/test/IsolatedLoggingRule.java [hunk 2]
```diff
@@ -132,4 +133,13 @@ private Field getLoggerField(String name) {
...
 
+    private void ensureSerialRunner(Description description) {
+        RunWith runWithAnnotation = description.getTestClass().getAnnotation(RunWith.class);
+        Class runnerClass = runWithAnnotation == null ? null : runWithAnnotation.value();
+        if (runnerClass == null || !HazelcastSerialClassRunner.class.isAssignableFrom(runnerClass)) {
+            throw new IllegalStateException("Isolated logging may be achieved only when running a test using a serial runner. Use"
+                    + " HazelcastSerialClassRunner or its subclass to run this test.");
+        }
+    }
+
```
Comment (blazember) [inline hazelcast/src/test/java/com/hazelcast/test/IsolatedLoggingRule.java]: :+1:

**Pull request merged**